### PR TITLE
add an AJAX behavior to transfer client side file information to ser…

### PR DESCRIPTION
**Pull request by reiern70**
Add an AJAX behavior to transfer client side file information to server side when a file is selected for upload. Now in AJAX upload example, when user selects a file, server side gets notified... with no need to submit file

image

IMHO this could be useful for instance to enable/disable some actions depending on file size and so on.

### CODE REVIEW

**Comment by solomax** [comment]
Maybe it worth JIRA and the line in CHANGELOG? :)

**Comment by reiern70** [comment]
> Maybe it worth JIRA and the line in CHANGELOG? :)

Yes. I firsts wanted to some feedback on wether this would be a good addition or not. In last case I will just roll such a class in our private project (as we need this functionlaity)

**Comment by svenmeier** [comment]
Nice, I like it.

**Comment by reiern70** [comment]
> Nice, I like it.

Thanks. I will create an ISSUE for it

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java [hunk 1]
```diff
@@ -47,6 +56,73 @@
...
+				throw new WicketRuntimeException("Behavior " + getClass().getName()
+						+ " can only be added to an instance of a FileUploadField");
+			}
+		}
+
+		@Override
+		protected void onUpdate(AjaxRequestTarget target)
+		{
+			Request request = RequestCycle.get().getRequest();
+			String fileName = request.getRequestParameters().getParameterValue("fileName").toString();
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:84]: nitpick: cache `request.getRequestParameters()` as a local variable

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java [hunk 2]
```diff
@@ -47,6 +56,73 @@
...
+			Long fileSize = request.getRequestParameters().getParameterValue("fileSize").toLong(0);
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:85]: s/Long/long/

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java [hunk 3]
```diff
@@ -47,6 +56,73 @@
...
+		/**
+		 * Called when a file, at client side is selected.
+		 *
+		 * @param target The {@link org.apache.wicket.ajax.AjaxRequestTarget}
+		 * @param fileName The client file name
+		 * @param fileSize The client file size
+		 * @param lastModified The {@link java.util.Date} when file was last modified.
+		 * @param mimeType The  MIME type of file.
+		 */
+		protected abstract void onFileSelected(AjaxRequestTarget target, String fileName, Long fileSize,
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:100]: s/Long/long/

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java [hunk 4]
```diff
@@ -47,6 +56,73 @@
...
+			{
+				@Override
+				public CharSequence getPrecondition(Component component)
+				{
+					return "if (this.files) {  window."+component.getMarkupId()+"_file = this.files[0]; return true; } " +
+							"else return false;";
+				}
+			});
+
+			String var = "window."+getComponent().getMarkupId()+"_file";
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:116]: Does this compile in Java11 ? 
`var` is a keyword in Java 10+. I think you will need to rename it.

**Review by martin-g — COMMENTED** [review]
I think this could be extended to support `multiple` attribute of `<input type="file"../>`

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java (hunk 4 above)
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:116]: image

yes.

**Comment by reiern70** [comment]
it seems I need to be added as member to apache in github so that I can commit directly to wicket instead of my fork. Can I do that directly or do I need to request that from infra?

**Comment by martin-g** [comment]
Visit https://gitbox.apache.org/setup/

**Comment by reiern70** [comment]
> I think this could be extended to support `multiple` attribute of `<input type="file"../>`

Yes. I will probably do that in another PR.

**Comment by reiern70** [comment]
> Visit https://gitbox.apache.org/setup/

Many thanks! Now it seems to be working for me.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java (hunk 1 above)
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:84]: fixed.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java (hunk 3 above)
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:100]: fixed

**Comment by reiern70** [comment]
Closing because I have created a new PR/branch with corrections and proper issue number

**Pull request closed without merging**